File: python/circular-buffer/circular_buffer.py
class BufferFullException(Exception):
    pass


class BufferEmptyException(Exception):
    pass


class CircularBuffer(object):
    def __init__(self, capacity):
        self.buffer = [''] * capacity
        self.capacity = capacity
        self.oldest_index = 0
        self.element_count = 0

    def read(self):
        if self.element_count == 0:
            raise BufferEmptyException('buffer is empty')
        res = self.buffer[self.oldest_index]
        self.oldest_index  = (self.oldest_index + 1) % self.capacity
        self.element_count -= 1
        return res


    def write(self, data):
        if self.element_count == self.capacity:
            raise BufferFullException('buffer is full')
        self.buffer[(self.oldest_index + self.element_count) % self.capacity] = data
        self.element_count += 1

    def overwrite(self, data):
        if self.element_count < self.capacity:
            self.write(data)
        else:
            self.buffer[self.oldest_index] = data
            self.oldest_index  = (self.oldest_index + 1) % self.capacity

    def clear(self):
        self.buffer = [''] * self.capacity
        self.oldest_index = 0
        self.element_count = 0

File: csharp/circular-buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;

public class CircularBuffer<T>
{
    int capacity;
    Queue<T> buffer;

    public CircularBuffer(int capacity)
    {
        this.capacity = capacity;
        buffer = new Queue<T>();
    }

    public T Read()
    {
        if (buffer.Count == 0) {
            throw new InvalidOperationException("buffer is empty");
        }

        return buffer.Dequeue();
    }

    public void Write(T value)
    {
        if (buffer.Count == capacity) {
            throw new InvalidOperationException("buffer is full");
        }
        buffer.Enqueue(value);
    }

    public void Overwrite(T value)
    {
        if (buffer.Count == capacity) {
            buffer.Dequeue();
        }
        Write(value);
    }

    public void Clear()
    {
        buffer.Clear();
    }
}